common/src/main/java/io/netty/util/concurrent/SimpleConcurrencyLimit.java

```diff
@@ -0,0 +1,197 @@
+                        public void run() {
+                            if (pendingHandler.isTimeoutFutureDone()) {
+                                return;
+                            }
+
+                            pendingHandler.handler.permitAcquisitionTimedOut();
+                        }
+                    }, timeout, unit);
+                    pendingHandler.setTimeoutFuture(timeoutFuture);
+                }
+
+                pendingHandlers.add(pendingHandler);
+                return;
+            }
+
+            final int nextPermitsInUse = currentPermitsInUse + 1;
+            if (permitsInUseUpdater.compareAndSet(this, currentPermitsInUse, nextPermitsInUse)) {
+                if (executor.inEventLoop()) {
+                    handler.permitAcquired(releaser);
+                } else {
+                    executor.execute(new Runnable() {
+                        @Override
+                        public void run() {
+                            handler.permitAcquired(releaser);
+                        }
+                    });
+                }
+                return;
+            }
+        }
+    }
+
+    private void acquire(final PendingHandler pendingHandler) {
+        for (;;) {
+            final int currentPermitsInUse = permitsInUse;
+            if (currentPermitsInUse == maxPermits) {
+                pendingHandlers.add(pendingHandler);
+                return;
+            }
+
+            final int nextPermitsInUse = currentPermitsInUse + 1;
+            if (permitsInUseUpdater.compareAndSet(this, currentPermitsInUse, nextPermitsInUse)) {
+                final EventExecutor executor = pendingHandler.executor;
+                if (executor.inEventLoop()) {
+                    pendingHandler.notifyAcquisition(releaser);
+                } else {
+                    executor.execute(new Runnable() {
+                        @Override
+                        public void run() {
+                            pendingHandler.notifyAcquisition(releaser);
+                        }
+                    });
+                }
+                return;
+            }
+        }
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName() +
+               "{maxPermits=" + maxPermits +
+               ", permitsInUse=" + permitsInUse +
+               '}';
+    }
+
+    private class Releaser implements Runnable {
+        @Override
+        public void run() {
+            for (;;) {
+                final int currentPermitInUse = permitsInUse;
+                final int nextPermitInUse = currentPermitInUse - 1;
+                assert nextPermitInUse >= 0 : "Released more than acquired";
+                if (permitsInUseUpdater.compareAndSet(SimpleConcurrencyLimit.this,
+                                                      currentPermitInUse, nextPermitInUse)) {
+                    final PendingHandler pendingHandler = pendingHandlers.poll();
```

Comment:
If the pending handler is non-null, we could in theory "hand over" our existing permit, and skip the decrement/increment dance.